File: 175 Edit Distance.py
from sys import stdin
import sys
sys.setrecursionlimit(10**7) 
def editDistance(word1, word2) :
    n,m = len(word1),len(word2)
    l = [[i for i in range(n+1)] for i in range(m+1)]
    
    for i in range(m+1):
        l[i][0] = i
    
    for i in range(m):
        for j in range(n):
            if word1[j]==word2[i]:
                l[i+1][j+1] = l[i][j]
            else:
                l[i+1][j+1] = min(l[i][j],l[i+1][j],l[i][j+1])+1

    return l[-1][-1]
    
#taking inpit using fast I/O
def takeInput() :
    str1=input()
    str2=input()

    return str1, str2


#main
str1, str2  = takeInput()
print(editDistance(str1, str2))

File: 175 Edit Distance.cpp
#include <bits/stdc++.h>
int dp[501][501];
int f(int i,int j,string word1,string word2)
{
    int n=word1.size(),m=word2.size();
    if (j==m)
        return n-i;
    if (i==n)
        return m-j;
    if (dp[i][j]!=-1)
        return dp[i][j];
    
    int ans = INT_MAX;
    
    if (word1[i]==word2[j])
        ans = min(ans,f(i+1,j+1,word1,word2));
    else
    {
        ans = min(ans,1+f(i,j+1,word1,word2));
        ans = min(ans,1+f(i+1,j+1,word1,word2));
        ans = min(ans,1+f(i+1,j,word1,word2));
    }
    return dp[i][j] = ans;
}

int editDistance(string word1, string word2) {
    
    memset(dp,-1,sizeof dp);
    return f(0,0,word1,word2);
}
